`cargo add`: Infer crate name from path/git repo

mseri opened the pull request:
This is an attempt to address issue #14. 
There are still few FIXME in the code (in part due to me waiting for the hyper PR to be merged).

killercup commented on README.md:
```diff
@@ -81,7 +83,9 @@ Options:
-Add a dependency to a Cargo.toml manifest file.
+Add a dependency to a Cargo.toml manifest file.  If <crate> is a github or gitlab repository url, or 
+a local path, `cargo add` will try to authomaticall get the crate name and authomatically set the 
+appropriate `--git` or `--path`. 
```
(Tiny typo: It's spelled 'automatically'. Also, please write URL in uppercase.)

killercup commented:
Thank you!

I just had a quick look at this—It looks good so far (I added a comment about a typo), but I haven't tested it locally. This changes cargo-add quick significantly, so I would like to have some more tests for this.

> There are still few FIXME in the code (in part due to me waiting for the hyper PR to be merged).

I'm not sure this PR (#67) will be merged. but I would like to refactor the fetch code in the future. No need to wait for this, though.

Regarding your `Option` FIXMEs: I would very much like to use `Result`s for all error handling!`Option` should only indicate missing values, never errors. Feel free to refactor any code that doesn't follow this rule. :smile: (Also, quick-error is a pretty easy way to create comprehensive error `enum`s.)

I'll be on vacation the next weeks, I'll probably check in every few days, though.

mseri commented:
Thanks for the comments! 

I will probably not be able to do much until Tuesday, I will push the changes concerning Option/Result as soon as possible. And then I will implement some new tests, do you have some that you would particularly like to see?

mseri commented:
I have cleaned the code, replaced Option to Result, and modified how I deal with paths -- although I am not convinced that I am doing it in the best way (I have left a FIXME there). 

I also added tests concerning success and failure. If you think that we should add more, just let me know what you believe it's missing. Setting the environment variable "NO_REMOTE_CARGO_TEST" allows to skip the new tests requiring a working network connection.

This version seems to work on my computer using stable. I don't understand where the SIGSEV in unstable comes from. Do you have any guess?

mseri commented:
@killecup any comment on this PR? I am using it since the latest commit and had no issue yet

mseri commented:
@killercup any comment on this PR? I am using it since the latest commit and had no issue yet

killercup commented:
Sorry for the radio silence, I didn't have as much time as I had hoped…

I'm not sure setting a `NO_REMOTE_CARGO_TEST` environment variable is the best way to do this. Rust's build-in testing framework allows you to mark tests as `#[ignore]`, and later only run those with `cargo test -- --ignored`. Using this, we can easily separate tests using external APIs. Another option would be to add a `test-external-apis` cargo feature. Both are rather self-explainatory—and opt-in.

> I don't understand where the SIGSEV in travis-unstable comes from. Do you have any guess?

No idea. Don't worry about it, I'll probably bump the nightly version later anyway to get a newer clippy.

killercup commented on tests/fixtures/add/local/Cargo.toml:
```diff
@@ -0,0 +1,7 @@
+[package]
```
Can you rename this to `Cargo.toml.sample`? We renamed all the other `Cargo.toml` files in the `fixtures` directory because it was confusing `cargo install`. Thanks!

mseri commented:
No problem. 

I am going to add a cargo feature then.
The SIGSEV was due to libcurl. It has been enough to bump the version to get rid of it.

killercup commented:
Cool, thank you! Let's land this PR now :smile:

@homu r+

homu commented:
:pushpin: Commit d227f01 has been approved by `killercup`

<!-- @homu r=killercup d227f01eb750ece803c017178e0ffc81f38c2a4d -->

homu commented:
:zap: Test exempted - [status](https://travis-ci.org/killercup/cargo-edit/builds/121937023)
